### PR TITLE
Simplify WDL Toil job graphs

### LINKED ISSUE #4465: Refactor WDL interpreter to run multiple WDL workflow nodes per Toil job
The other half of #4388 is to divide up the WDL workflow node graph into components that can be run as units on the leader node, so we don't need so many distinct Toil jobs.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1328)
┆Issue Number: TOIL-1328

### PR DESCRIPTION
**Pull request by adamnovak**
This will fix #4465 by allowing multiple WDL workflow nodes to (at least sometimes) live in one Toil job.

It also removes some Toil jobs that don't really need to be separate jobs, by baking the transformations they would do into the `WDLBaseJob`.

I ran a version of the conformance tests (2b91928600bfb1397c190a9900ac323db5111d1b) against this and I got:

Before:

```
64 tests run, 37 succeeded, 27 failed, 6 skipped
    Failures: 19,20,27,31,41,42,43,44,45,46,47,48,49,50,51,52,54,55,56,60,61,63,64,65,66,68,69
DOCKER_HOST=unix:///Users/anovak/.docker/run/docker.sock python3 run.py    1.  334.19s user 104.37s system 297% cpu 2:27.39 total
```

After:

```
64 tests run, 37 succeeded, 27 failed, 6 skipped
	Failures: 19,20,27,31,41,42,43,44,45,46,47,48,49,50,51,52,54,55,56,60,61,63,64,65,66,68,69
DOCKER_HOST=unix:///Users/anovak/.docker/run/docker.sock python3 run.py    1.  152.97s user 49.88s system 279% cpu 1:12.58 total
```

So it's maybe twice as fast and no *less* conformant.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * WDL interpreter job postprocessing operations can now get attached to the nearest job with actual content instead of needing to be separate whole jobs
 * Multiple WDL declarations can now be evaluated in one job

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
Looks like this failed the Giraffe WDL test with:

```
[2023-07-06T22:13:22+0000] [MainThread] [W] [toil.leader] Log from job "kind-WDLWorkflowNodeListJob/instance-p3kiwm04" follows:
=========>
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.worker] ---TOIL WORKER OUTPUT LOG---
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil] Running Toil version 5.12.0a1-b6cfd132e2cba6a361e00dbcecd674483562586c on host runner-svulqt5t-project-3-concurrent-12bgtdt.
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.worker] Working on job 'WDLWorkflowNodeListJob' decl-reference_index_file+ kind-WDLWorkflowNodeListJob/instance-p3kiwm04 v1
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.worker] Loaded body Job('WDLWorkflowNodeListJob' decl-reference_index_file+ kind-WDLWorkflowNodeListJob/instance-p3kiwm04 v1) from description 'WDLWorkflowNodeListJob' decl-reference_index_file+ kind-WDLWorkflowNodeListJob/instance-p3kiwm04 v1
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.wdl.wdltoil] Setting reference_index_file to select_first([REFERENCE_INDEX_FILE, indexReference.reference_index_file])
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.wdl.wdltoil] Setting reference_dict_file to select_first([REFERENCE_DICT_FILE, indexReference.reference_dict_file])
	[2023-07-06T22:13:21+0000] [MainThread] [I] [toil.job] Saving graph of 1 jobs, 1 non-service, 0 new
	Traceback (most recent call last):
	  File "/builds/databiosphere/toil/src/toil/worker.py", line 403, in workerScript
	    job._runner(jobGraph=None, jobStore=jobStore, fileStore=fileStore, defer=defer)
	  File "/builds/databiosphere/toil/src/toil/job.py", line 2782, in _runner
	    self._saveJobGraph(jobStore, saveSelf=False, returnValues=returnValues)
	  File "/builds/databiosphere/toil/src/toil/job.py", line 2572, in _saveJobGraph
	    self._fulfillPromises(returnValues, jobStore)
	  File "/builds/databiosphere/toil/src/toil/job.py", line 2318, in _fulfillPromises
	    pickle.dump(promisedValue, fileHandle, pickle.HIGHEST_PROTOCOL)
	RecursionError: maximum recursion depth exceeded while calling a Python object
	[2023-07-06T22:13:21+0000] [MainThread] [E] [toil.worker] Exiting the worker because of a failed job on host runner-svulqt5t-project-3-concurrent-12bgtdt
<=========
```

Maybe I managed to override the recursion limit bump code to not get called?

File: src/toil/wdl/wdltoil.py [hunk 1]
```diff
@@ -885,6 +890,11 @@ def __init__(self, **kwargs: Any) -> None:
         # TODO: Make sure C-level stack size is also big enough for this.
         sys.setrecursionlimit(10000)
 
+        # We need an ordered list of postprocessing steps to apply, because we
+        # may ahve coalesced postprocessing steps deferred by several levels of
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:894]: `ahve`

File: src/toil/wdl/wdltoil.py [hunk 2]
```diff
@@ -885,6 +890,11 @@ def __init__(self, **kwargs: Any) -> None:
...
+        # jobs returing other jobs' promised RVs.
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:895]: `returing`

File: src/toil/wdl/wdltoil.py [hunk 3]
```diff
@@ -1383,43 +1521,140 @@ def __init__(self, prev_node_results: Sequence[Promised[WDLBindings]], underlay:
...
+
+        Produces dependencies after resolving gathers and internal-to-section
+        dependencies, on nodes that are also in this graph.
+        """
+
+        # We need to make sure to bubble up dependencies from inside sections.
+        # A conditional might only appear to depend on the variables in the
+        # conditional expression, but its body can depend on other stuff, and
+        # we need to make sure that that stuff has finished and updated the
+        # environment before the conditional body runs. TODO: This is because
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:1589]: We should make a ticket for this.

File: src/toil/wdl/wdltoil.py [hunk 4]
```diff
@@ -1434,7 +1669,76 @@ def __init__(self, namespace: str, **kwargs: Any) -> None:
         super().__init__(**kwargs)
         self._namespace = namespace
 
-    def create_subgraph(self, nodes: Sequence[WDL.Tree.WorkflowNode], gather_nodes: Sequence[WDL.Tree.Gather], environment: WDLBindings, local_environment: Optional[WDLBindings] = None) -> Job:
+    def coalesce_nodes(self, order: List[str], section_graph: WDLWorkflowGraph) -> List[List[str]]:
+        """
+        Given a topological order of WDL workflow node IDs, produce a list of
+        lists of IDs, still in topological order, where each list of IDs can be
+        run under a single Toil job.
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:1676]: Well stated.

**Review by DailyDreaming — COMMENTED** [review]
This looks well done.  I like the organization and the comments are invaluable.  Mostly minor comments, except I wonder if there should be a test for if something like `coalesce_nodes` running IDs together that we would expect it to run together.

**Review by DailyDreaming — APPROVED** [review]
Looks great!  Thanks for adding the (sub)tests.

**Pull request merged**